### PR TITLE
#1749 plot_matching overlays robust matching

### PR DESCRIPTION
**Pull request by ahundt**
This currently illustrates the bug present in image warping, showing an inconsistency in the resulting transformation matrices.
This change also improves the example to be more useful than before, so at least parts of it should be kept once the underlying bug is fixed.

### CODE REVIEW

**Comment by stefanv** [comment]
@ahojnnes Do you have a few minutes to look at this?

File: doc/examples/plot_matching.py [hunk 1]
```diff
@@ -109,8 +135,8 @@ def match_corner(coord, window_ext=5):
 model.estimate(src, dst)
 
 # robustly estimate affine transform model with RANSAC
-model_robust, inliers = ransac((src, dst), AffineTransform, min_samples=3,
-                               residual_threshold=2, max_trials=100)
+model_robust, inliers = ransac((src, dst), AffineTransform, min_samples=8,
```
Comment (ahojnnes) [inline doc/examples/plot_matching.py]: 8 rather than 3 samples defeats the whole purpose of RANSAC.

File: doc/examples/plot_matching.py [hunk 2]
```diff
@@ -120,7 +146,20 @@ def match_corner(coord, window_ext=5):
 print(model_robust.scale, model_robust.translation, model_robust.rotation)
 
 # visualize correspondence
-fig, ax = plt.subplots(nrows=2, ncols=1)
+fig, ax = plt.subplots(nrows=4, ncols=1)
+
+def warp_and_combine(img1,img2,xform):
+    img1_warped = warp(img1,xform)
```
Comment (ahojnnes) [inline doc/examples/plot_matching.py]: I didn't look carefully, but you might need to swap the 0-th and 1-th dimension as the warp function expects y, x and the affine transform might return x, y.

File: doc/examples/plot_matching.py [hunk 3]
```diff
@@ -64,44 +84,50 @@
...
 
-def match_corner(coord, window_ext=5):
-    r, c = np.round(coord).astype(np.intp)
-    window_orig = img_orig[r-window_ext:r+window_ext+1,
-                           c-window_ext:c+window_ext+1, :]
 
-    # weight pixels depending on distance to center pixel
-    weights = gaussian_weights(window_ext, 3)
-    weights = np.dstack((weights, weights, weights))
+def match_corner(coord_list1,coord_list2,img1,img2, patch_radius=5):#     # find correspondences using simple weighted sum of squared differences
```
Comment (ahojnnes) [inline doc/examples/plot_matching.py]: What's the purpose of this change?

**Comment by ahojnnes** [comment]
I will try to run it later tonight to see what's going on.

File: doc/examples/plot_matching.py (hunk 3 above)
Comment (ahundt) [inline doc/examples/plot_matching.py]: cleanup explaining what's going on and making the pieces of the code here easier to split apart

File: doc/examples/plot_matching.py (hunk 1 above)
Comment (ahundt) [inline doc/examples/plot_matching.py]: that's fine to change back, was just experimenting

File: doc/examples/plot_matching.py (hunk 2 above)
Comment (ahundt) [inline doc/examples/plot_matching.py]: I suggest that discrepancy either be resolved by API changes (what I think to be ideal) or at minimum documentation of it in the APIs. Of course either way the example here should be updated to work correctly as well.

**Comment by ahundt** [comment]
Any luck with this?

**Comment by stefanv** [comment]
I am traveling for the next week, so I'll let Johannes handle this.

**Comment by ahojnnes** [comment]
@ahundt Sorry for the late reply. I have been very busy lately. I'll try to have a look later today.

**Comment by ahojnnes** [comment]
In fact, it turns out I was right :-)

The AffineTransform is estimated using coordinates as "Y, X", but the warp function expects the transformation in the format: "X, Y".  Simply add a
```
print(model_robust.params)
model_robust.params[[0, 1], :] = model_robust.params[[1, 0], :]
print(model_robust.params)
```

before plt.gray() and you should be fine. It is explained in the documentation, but I try to highlight that fact a little more. Let me know if that fixes your problems. Thanks for your feedback!

**Comment by ahojnnes** [comment]
I created a new issue for this: https://github.com/scikit-image/scikit-image/issues/1789

**Comment by ahundt** [comment]
That work great! I updated the example to use the fix. Can my improvements to the example be merged?

**Comment by ahundt** [comment]
How can I fix the travis CI issue? I'm actually using 2.7 so I'm not sure why that's what is currently failing

**Comment by ahundt** [comment]
@ahojnnes It looks like 2.7 is failing due to an issue unrelated to the code here.

**Comment by ahojnnes** [comment]
Since it is confirmed that this issue is not a bug but rather an inconsistency in the implementation, which is documented in another issue, I am closing this issue here. @ahundt Please, let me know if you think otherwise.

**Comment by ahundt** [comment]
@ahojnnes The other part of the issue that still stands open is that I would like to get the example improvements I added merged! I'd appreciate if this could be kept open so we can make that happen.

**Comment by ahundt** [comment]
@ahojnnes I would like to contribute the improved example, that's why I submitted this pull request. Please consider reopening this!

**Comment by ahojnnes** [comment]
@ahundt Could you please elaborate in how your modified example improves over the current one? Thanks!

**Comment by ahundt** [comment]
Here are some improvements:
- more illustrative visualization
  - accuracy of the detected warp
  - how the two images compare

ski_plot_matching_combine_and_warp

- correctly apply a warp (nontrivial due to xy-yx swap)
- overlay two images
- organizes the key points in the same manner that `corner_peaks()` and `plot_matches()`
   - this means it is easier to take the new example and integrate it with other use cases since the data is organized more consistently
- more robust to changes in checker scale, rotation and translation 
    - the previous version fails very badly with even minor changes, the new one is better but not great
- easy to try with your own image pair, just set `original=False` and rename `imageName` and `imageName2` to the name of your own files.

**Comment by stefanv** [comment]
@ahundt I haven't been following closely, but why are the estimates not robust?

Thanks for helping us to improve scikit-image!

**Comment by ahundt** [comment]
@stefanv If you change the theta and x,y offset to various reasonable values the match will fail, or it will incorrectly find a simple translation when the image is in reality a combination of translation and rotation. Sorry but I didn't save any examples of that to illustrate.

I added the x,y position difference as a weighting factor in the match. While it won't always work it seems appropriate in this case. If even better matches are desired I guess a better feature detector would be needed such as ORB rather than simple ssd.

**Comment by stefanv** [comment]
I guess I just wanted to double-check if there is anything we are doing that is fundamentally wrong and should be improved, or whether the problem is just badly posed.  If the former, I will have to dig in deeper to understand better.

**Comment by ahundt** [comment]
No fundamental problems, the methods you use match up with what I understand to be correct. 

I'm just reorganizing a bit and adding some details, because when I went to learn from the example the next things I tried to do didn't really have much documentation, so I added the most likely tasks (i.e. warping, which initially appeared broken) to the example and submitted the changes since I think they'll be useful to others.

**Comment by ahojnnes** [comment]
The main source of instability of the code is the matching and the highly ambiguous image corners. The image overlay would definitely be a nice addition to the example. I don’t mean to be rude, but the other changes don’t make much sense in my opinion: 1) Removing the gradient from the image makes the patches even more ambiguous. 2) Changing the number of RANSAC samples from 3 to 8 is basically defeating the purpose of RANSAC - the number of samples should always be minimal (with very rare exceptions). 3) The matching is basically equivalent, you just pulled the loop over the corners into the function.

I am happy to review a PR with the image overlay as an addition. Would you mind creating a new PR for this? Please, also make sure to conform to our coding style guidelines (PEP008). Thanks!

**Comment by ahundt** [comment]
I'll try and fix some things up and make a new pull request when I have time. 

I unfortunately need to disagree with you on the ransac point. While 3 points is the minimum for a match, It is not unlikely for 3 random points to line up nicely, and incorrectly, as a simple translation when the real change is both a translation and a rotation. 

In fact, encountering this issue is the very reason I made this change, so it isn't theoretical.

Requiring more points to match up consistently means that the probability of an incorrect but consistent estimate is correspondingly lower, particularly when the data provides enough correct results that you won't have trouble finding 8 points for which the transform is consistent.

**Comment by stefanv** [comment]
That's an interesting argument, and one we should consider more closely.  Even reasoning about this i.t.o. a line model, one is tempted to think that a solution can be found much more rapidly if each fit is a least squares fit; that is, *if* you are lucky enough not to accidentally draw outliers.  There are two extremes: draw as few samples as possible, minimizing the chance of drawing outliers, or drawing as many data points as possible (least squares).  Clearly, we want to stay close to the lower boundary, but the question is how close.

From http://homepages.inf.ed.ac.uk/rbf/CVonline/LOCAL_COPIES/CANTZLER2/ransac.pdf:

"Drawing more than the minimal number of sample points is inefficient, since the probability of selecting a sample consisting only of inlying data points (i.e. all data points belonging to the same model), that gives a good estimate and at random, decreases with respect to the increasing sample size.  Thus the minimal sample set maximizes the probability of selecting a set of inliers from which later a good estimate will be computed."

Perhaps we should consider another question: do we have the correct metric for inliers?  Because that is what determines whether a model is *valid* or not, and if a pathological fit is considered valid, clearly we have a problem.

**Comment by ahundt** [comment]
Yes, that makes perfect sense. I wish I saved one of the so-called pathological matches because then it would be easier to see exactly how it is failing. I'll do my best from memory.

If the first match is an excellent pathological match, then it is conceivable that continued attempts at matching where you land on accurate matches are thrown away, even if you would build a much larger consensus had you landed on an accurate match. If I remember correctly, I'd get maybe 3-4 points on a pathological match, while if I re-ran and got lucky, I'd get a dozen matches on accurate points.

I think the easiest way to reproduce that would be a translation of 1 small square with no rotation. However I may not remember that  right.

**Comment by stefanv** [comment]
If you can reproduce this, that would provide us with an excellent test bed.

**Comment by ahundt** [comment]
others have reproduced this in the original threadhttps://github.com/scikit-image/scikit-image/issues/1749#issuecomment-458613786
